### PR TITLE
feat(upgrade): support v2 volume upgrade

### PR DESCRIPTION
**Pull request by derekbit**
#### Which issue(s) this PR fixes:

Issue longhorn/longhorn#6001

### CODE REVIEW

**Review by DamiaSan — APPROVED** [review]
LGTM.
Just a little comment: some functions, like for example Engine.Create, has become quite long and complex, maybe in the future we could make a refactor to simplify the code.

**Comment by derekbit** [comment]
@DamiaSan @shuo-wu 
Can you help review it again? Thank you.

The modifications to `engine.SwitchOverTarget()` focus on error handling. If the SPDK engine fails to switch over the target, it attempts to reconnect to the original target. This allows the control plane (longhorn-manager) to determine the next step: either abandon the switchover or retry the process.

File: pkg/spdk/engine.go [hunk 1]
```diff
@@ -1535,3 +1685,329 @@ func (e *Engine) RestoreStatus() (*spdkrpc.RestoreStatusResponse, error) {
...
+		return err
+	}
+
+	return nil
+}
+
+func (e *Engine) IsTargetDisconnected() (bool, error) {
+	initiator, err := nvme.NewInitiator(e.VolumeName, e.Nqn, nvme.HostProc)
+	if err != nil {
+		return false, errors.Wrapf(err, "failed to create initiator for engine %s target switchover", e.Name)
```
[the remark this answers is not in the document]
Comment (derekbit) [reply pkg/spdk/engine.go]: Updated. Thank you.

File: pkg/spdk/engine.go [hunk 2]
```diff
@@ -1535,3 +1685,329 @@ func (e *Engine) RestoreStatus() (*spdkrpc.RestoreStatusResponse, error) {
...
+	if err != nil {
+		return false, errors.Wrapf(err, "failed to check if engine %s is suspended", e.Name)
+	}
+	if !suspended {
+		return false, fmt.Errorf("engine %s must be suspended before target switchover", e.Name)
+	}
+
+	if err := initiator.LoadNVMeDeviceInfo(initiator.TransportAddress, initiator.TransportServiceID, initiator.SubsystemNQN); err != nil {
+		if !nvme.IsValidNvmeDeviceNotFound(err) {
+			return false, errors.Wrapf(err, "failed to load NVMe device info for engine %s target switchover", e.Name)
```
[the remark this answers is not in the document]
Comment (derekbit) [reply pkg/spdk/engine.go]: Updated. Thank you.

**Review by DamiaSan — APPROVED** [review]

**Pull request merged**